zebra-state/src/service/check/nullifier.rs

```diff
@@ -56,40 +56,33 @@ pub(crate) fn no_duplicates_in_finalized_chain(
     Ok(())
 }
 
-/// Reject double-spends of nullifiers:
-/// - one from this [`Transaction`], and the other already committed to the
-///   provided [`ZebraDb`].
-///
-/// # Consensus
-///
-/// > A nullifier MUST NOT repeat either within a transaction,
-/// > or across transactions in a valid blockchain.
-/// > Sprout and Sapling and Orchard nullifiers are considered disjoint,
-/// > even if they have the same bit pattern.
-///
-/// <https://zips.z.cash/protocol/protocol.pdf#nullifierset>
-// TODO: impl nullifier getters for a trait and use a generic type to unify
-//       this fn with the one for `PreparedBlock`
-#[tracing::instrument(skip_all)]
-fn tx_no_duplicates_in_finalized_chain(
-    finalized_chain: &ZebraDb,
-    transaction: &Arc<Transaction>,
-) -> Result<(), ValidateContextError> {
-    for nullifier in transaction.sprout_nullifiers() {
-        if finalized_chain.contains_sprout_nullifier(nullifier) {
-            Err(nullifier.duplicate_nullifier_error(true))?;
-        }
-    }
-
-    for nullifier in transaction.sapling_nullifiers() {
-        if finalized_chain.contains_sapling_nullifier(nullifier) {
-            Err(nullifier.duplicate_nullifier_error(true))?;
+/// Returns a Ok(()) if the predicate(s) return false for every revealed nullifier
+/// Returns Err(DuplicateNullifierError) for the first nullifier that matches the predicate(s)
+fn find_duplicate_nullifier<'a, NullifierT, FinalizedStateContainsFn, NonFinalizedStateContainsFn>(
+    revealed_nullifiers: impl IntoIterator<Item = &'a NullifierT>,
+    finalized_state_contains: FinalizedStateContainsFn,
+    non_finalized_state_contains: Option<NonFinalizedStateContainsFn>,
+) -> Result<(), ValidateContextError>
+where
+    NullifierT: DuplicateNullifierError + 'a,
+    FinalizedStateContainsFn: Fn(&'a NullifierT) -> bool,
+    NonFinalizedStateContainsFn: Fn(&'a NullifierT) -> bool,
+{
+    if let Some(non_finalized_state_contains) = non_finalized_state_contains {
+        for nullifier in revealed_nullifiers {
+            if non_finalized_state_contains(nullifier) {
+                Err(nullifier.duplicate_nullifier_error(false))?
+            }
+
+            if finalized_state_contains(nullifier) {
+                Err(nullifier.duplicate_nullifier_error(true))?
+            }
         }
-    }
-
-    for nullifier in transaction.orchard_nullifiers() {
-        if finalized_chain.contains_orchard_nullifier(nullifier) {
-            Err(nullifier.duplicate_nullifier_error(true))?;
+    } else {
+        for nullifier in revealed_nullifiers {
+            if finalized_state_contains(nullifier) {
+                Err(nullifier.duplicate_nullifier_error(true))?
+            }
         }
```

Comment:
Nitpick: I'm not sure why we're duplicating the finalized chain code, can we put the `if let` inside the loop instead?

Reply:
Updated in [`7ac6684`](https://github.com/ZcashFoundation/zebra/pull/5717/commits/7ac668401e71143c0386e1f45316b6c4ce2f4a2e)